src/Interpreters/TreeRewriter.cpp

```diff
@@ -1375,7 +1378,27 @@ TreeRewriterResultPtr TreeRewriter::analyzeSelect(
     result.aggregates = getAggregates(query, *select_query);
     result.window_function_asts = getWindowFunctions(query, *select_query);
     result.expressions_with_window_function = getExpressionsWithWindowFunctions(query);
+
+    if (is_parameterized_view)
+    {
+        for (auto & column : result.source_columns)
+        {
+            std::string column_name = column.name;
+            std::string::size_type pos = 0u;
+            for (auto & parameter : parameter_values)
+            {
+                if ((pos = column_name.find(parameter.first)) != std::string::npos)
+                {
+                    String parameter_name("_CAST(" + parameter.second + ", '" + column.type->getName() + "')");
```

Comment:
as understand it is for case `CREATE VIEW pv6 AS SELECT Price+{price:UInt64} ...`? So the cast is needed to understand result type?

Reply:
Here we actually replace the source_columns in result with CAST, the column without CAST is no longer referred to. translateQualifiedNames replaces the column in the query and here we update source_columns, which is used to get required_source_columns.

Reply:
ok, I see, so what you do here is probably the same reason as for this line of code for ordinary query parameters: https://github.com/ClickHouse/ClickHouse/blob/97e83eb53ce3efb655d63f30848ef540430c791a/src/Interpreters/ReplaceQueryParameterVisitor.cpp#L88 which also adds cast: https://github.com/ClickHouse/ClickHouse/blob/97e83eb53ce3efb655d63f30848ef540430c791a/src/Interpreters/addTypeConversionToAST.cpp#L23
but for you case you also use `replaceQueryParameterVisitor`, so the cast should be added there, so why here we also need to do cast here?

Reply:
So replaceQueryParameterVisitor replaces the query parameter and added the actual value, but it does the replacement in the actual query (not columns). 
In TreeRewriter, we get the actual columns plus(Price, _CAST(10, 'UInt64')) and match it to the columns in source which will be plus(Price, price) (thus the substitution to match in translateQualifiedNames). Then here finally we can replace source_columns with _CAST.
Hope that answers your question.

Reply:
Ok, so can we add this cast when we create the parametrised view itself, so we will always have this cast in source columns and will not need to add/remove it on certain time? it just scares me that in one place we need to remove it explicitly by scanning through column names, seems that we can as easily loose such place somewhere else.

Reply:
We need the value of the literal for the _CAST substitution, so cant do this when we create parameterized view